### PR TITLE
feat: `cool` cheatcode

### LINKED ISSUE #1340: `vm.cool` cheatcode
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (2865fa5 2022-04-14T21:44:28.244220110+00:00)

### What command(s) is the bug in?

forge test

### Operating System

Linux

### Describe the bug

With the introduction of the Berlin hardfork [EIP2929](https://eips.ethereum.org/EIPS/eip-2929) was implemented and went live on the Ethereum blockchain. This EIP introduced cost changes to storage and account accessing opcodes, including (but not limited to): `SLOAD`, `SSTORE`, `CALL`, `BALANCE` etc. Specifically these opcodes received an additional cost the first time they are used to access storage slots and/or accounts within a transaction. These costs are however not always properly accounted for in foundry solidity tests. Specifically addresses and storage slots seem to remain warm between calls within the context of a single test, falsely reducing the cost of calls and certain contract executions. This not only leads to inaccurate traces but gas usage dependent tests may also incorrectly fail / pass.

This bug may or may not be a feature. It is likely more useful / straightforward if accounts and storage slots do not remain warm between calls within a single test by default. Cheat codes may be added to purposefully keep accounts and storage slots warm or potentially even pre-warm to simulate transactions with access lists ([EIP2930](https://eips.ethereum.org/EIPS/eip-2930)).

EDIT(onbjerg): I am requesting a `vm.cool(address)` cheatcode to cool off warm accounts

### PR DESCRIPTION
**Pull request by ditto-eth**
## Motivation

Closes #1340.

Motivation is wanting to have more accurate and easier gas tests in foundry without needing to do hacks around functions/setUp. Adds a cheatcode to reset a warm account to cold and it's storage to be untouched/cold via `vm.cool(address)`

## Solution

> not familiar with the language/code so maybe some things are pretty wrong, would love some guidance on what to change or if something is missing!

it untouches the `Account.status` and clears all storage cache in the `Account` via `Account.storage`

also had some code relating to removing the `JournalEntry` for `JournalEntry::StorageChange` but it wasn't necessary for this to work in the tests i wrote - so i don't know if it's supposed to be cleared or not but it's not effecting how the gas is calculated at least in this basic test?

i'm not sure the best way to test other than via `gasleft` - instead of hardcoding gas values that may change it just asserts that the same calls have the same gas values and the warm call is less gas. i wrote a few more tests using sstore (20k for zero to non zero and 5k for nonzero to nonzero) but wasn't sure it was necessary

(didn't account for refunds if that is related, since gasleft doesn't do it)

see also: twitter [context](https://x.com/rakitadragan/status/1702047574075535562?s=20)

### CODE REVIEW

**Review by Evalir — CHANGES_REQUESTED** [review]
nice, this is quite clean and straightforward. Just a smol nit about moving the impl to a function, but this is good!

**Review by mattsse — APPROVED** [review]
tests look good.

tysm

File: crates/evm/src/executor/inspector/cheatcodes/env.rs [hunk 1]
```diff
@@ -387,6 +387,17 @@ pub fn apply<DB: DatabaseExt>(
             )?;
             ru256_to_u256(val).encode().into()
         }
+        HEVMCalls::Cool(inner) => {
+            ensure!(!is_potential_precompile(inner.0), "Load cannot be used on precompile addresses (N < 10). Please use an address bigger than 10 instead");
```
Comment (ditto-eth) [inline crates/evm/src/executor/inspector/cheatcodes/env.rs:391]: is this line needed actually?

**Review by Evalir — APPROVED** [review]
lgtm!

**Comment by Evalir** [comment]
let's also remember to update the book! @ditto-eth

**Comment by ditto-eth** [comment]
yep already have a commit haha!

File: crates/evm/src/executor/inspector/cheatcodes/env.rs [hunk 2]
```diff
@@ -318,6 +318,18 @@ fn add_breakpoint(state: &mut Cheatcodes, caller: Address, inner: &str, add: boo
...
+fn cool_account<DB: DatabaseExt>(data: &mut EVMData<'_, DB>, address: Address) -> Result {
+    if let Some(account) = data.journaled_state.state.get_mut(&h160_to_b160(address)) {
+        if account.is_touched() {
+            account.unmark_touch();
+        }
+        account.storage.clear();
+    }
+
+    Ok(Bytes::new())
+}
```
Comment (mds1) [inline crates/evm/src/executor/inspector/cheatcodes/env.rs:331]: I was reviewing this when I saw the forge-std PR and want to discuss a few things since this implementation feels wrong to me if I'm understanding it correctly. (See https://eips.ethereum.org/EIPS/eip-2929 for warm/cold specs)

First, there's an invariant that we must make sure is preserved: if a slot is warm, the address must also be warm. Consequently, if an address is cold, all it's slots must also be cold. I think the current implementation preserves this, but worth keeping in mind

Now, my understanding is the use cases for this cheat are primarily gas benchmarking, so you can do some setup then mark a contract and its slots as cold to get a worst-case estimate of transaction costs. Given that:

## Issue 1

IIUC, `account.unmark_touch()` marks an address cold, and `account.storage.clear()` clears all of it's storage to make the slots cold. (`testCool_SLOAD` has no assertions on the storage values remaining unchanged and I haven't looked deeper).

But slots don't have to be zero to be cold. At the start of a transaction, storage is nonzero in a contract but its slots are all cold. So if I'm correct about the existing behavior here, I'd say it's not desirable in many (all?) cases, given the use case given above. I'm not sure of a use case where you want to clear storage. The behavior here should be:

```solidity
// Marks the `target` address cold, and is a no-op if the address is already cold.
// All storage slots are are also made cold, but their values are preserved.
function coolAccount(address target) external;
```

## Issue 2

Less of an issue, more of a feature request so this is comprehensive, but we probably also want cheatcodes to cool certain slots, without cooling the address itself:

```solidity
struct StorageSlot {
  address target;
  bytes32 slot;
}

// Marks the specified `(target, slot)` pairs cold, and is a no-op if they are already cold.
// Reverts if the `target` has no code, since if there's no code it cannot have storage. 
// The values of storage slots are preserved.
function coolSlots(StorageSlot[] memory slots) external;
```

## Issue 3

Additionally, some addresses are warm at the start of a transaction:
- The `from` address
- The `to` address (or the address being created if it is a contract creation transaction)
- All precompiles

Therefore these addresses can never be cold, so we should make sure to restrict that accordingly.

For precompiles this is easy: revert if a precompile address is passed in.

For the `from` and `to` address it's complicated because of how forge works (each test is one giant transaction), but perhaps we document this behavior (i.e. that from and to are always warm) to make sure users use this cheat properly. The other option is to pretend all calls that are executed from the test contract are individual txs, and therefore treat the sender address (e.g. the default sender or `prank`/`broadcast` addr) AND the target of each call as always being warm. This might get complicated though.

Comment (ditto-eth) [reply crates/evm/src/executor/inspector/cheatcodes/env.rs:331]: wow, thanks for the precise feedback! I'm happy to work with you to get this right! Could I message you on telegram/twitter for further questions later?

for the precompile, I left check in earlier and removed it https://github.com/foundry-rs/foundry/pull/5830#discussion_r1326065936

Comment (ditto-eth) [reply crates/evm/src/executor/inspector/cheatcodes/env.rs:331]: k there's probably already way much to add here that's it's become an essay
but i'll make a few more notes (been back and forth thinking I had a better idea but maybe not).

i recovered the old code I had before this simpler PR and the other tests.
the merged code *does* have the issue that the storage slots are not changed since they are cleared/reset if you do a `sstore`. The gas calculations for sstore/sload are correct because the `account.storage` is empty and it does make sense. 

The problem seems to be that the state changes should still be there (be commited to the db?).
I tried another approach where I looped through `data.journaled_state.journal` to find each `JournalEntry::StorageChange` and then manually set each `account.storage`'s slot's to be `slot.previous_or_original_value = slot.present_value` which works to make sure it looks like the storage wasn't changed during the tx. But the gas values weren't what I expected.

(nz is non-zero, z is zero)
- a nz-nz sstore is 5k gas when slot is cold and that works for the first test when changing from 1 to 2.
- if I run `vm.cool` after, another nz-nz sstore is calculated to be 2.9k gas (which is 5k - 1200), so it means that the `vm.cool` didn't cool the slot, it just changed the value be the same.
- nz-nz without doing a `vm.cool` does calc a gas of 100, which is correct if the slot was already set earlier.

Doing more investigation I realized that in revm there is code that just checks the db.

https://github.com/bluealloy/revm/blob/d615514e09d7c7f5bf6a5d6d8b1a68c7aec59caf/crates/revm/src/journaled_state.rs#L624-L635

```rust
let load = match account.storage.entry(key) {
  Entry::Occupied(occ) => (occ.get().present_value, false), // returns is_cold is false
  Entry::Vacant(vac) => { .. } // is_cold is true?
}
```

this seems to show that if there is a `key` in `account.storage` that exists at all, revm thinks the `key` is warm? So it seems correct to clear it via `account.storage.clear()`, only problem being that the state changes that did happen need to persist. 

The code later calls `db.storage(address, key)?` - so there's another db that I assume stores the persistent storage? Make it's incorrect to change that though? Or is that similar to how forks work? Either way I didn't see any place to modify the `db`, other then maybe a tx commit?

so it's just looking like the way it's coded with `account.storage`, it needs to be cleared to be considered cool, but the sstores that were set before the `vm.cool` call still need to persist.

(meaning if there was a variable `slot0` that was already 1, set to 2, and then `vm.cool` happens, it should stay 2)
```solidity
// slot0 is already 1
// ..
slot0 = 2;
vm.cool(address(this));
assertEq(slot0, 2); // the merged pr currently resets the slot0 to 1
```

So the q is just how to preserve the changes in a slot's storage but make it cool, when the current `sload` code in revm sets `is_cold` if the key is in the `account.storage`

also, addressing @mds1's points, I think we were chatting on telegram that
`vm.cool(address target)` and `vm.cool(StorageSlot[] memory slots)`
should work fine if we overload the same name and different signatures.

**Comment by ditto-eth** [comment]
see https://github.com/foundry-rs/foundry/pull/5852 for my new PR
